### ISSUE TITLE
Args with a generic type will use the mother type, rather than a literal

### ISSUE THREAD

**Issue by gerhardadler** [python/mypy#17090]
**Bug Report**
Args with a generic type will use the mother type, rather than a literal

I noticed this issue while trying to chain together two lists with different types with `itertools.chain`.

**To Reproduce**

```python
from typing import TypeVar

T = TypeVar('T')

class M:
    pass


class C1(M):
    def my_method(self) -> None:
        pass

class C2(M):
    def my_method(self) -> None:
        pass


def test(*args: T) -> T:
    return args[0]

result = test(C1(), C2())
result.my_method()
```
https://mypy-play.net/?mypy=latest&python=3.12&flags=strict&gist=d78c6875b87a3040126e1926b8f7ba0c

**Expected Behavior**
I expected to be able to run `my_method`, as all arguments have it defined. Instead the mother type is used. This issue can be mitigated by extracting the arguments, giving them a literal type, then unpack them into the function:

```py
my_args: list[C1 | C2] = [C1(), C2()]
test(*my_args)
```

**Actual Behavior**
```
main.py:26: error: "M" has no attribute "my_method"  [attr-defined]
```

I believe this happens because untyped lists defined with items will use the mother type of all items. I don't think this is very intuitive for args.

Visual Studio Code's Pylance does not give an error in this situation, even with the strict flag.

**Your Environment**

- Mypy version used: 1.9.0
- Mypy command-line flags: no flags
- Mypy configuration options from `mypy.ini` (and other config files): no config
- Python version used: 3.9, 3.10, 3.11, 3.12

**Comment by erictraut** [comment]
This is another [join-v-union](https://github.com/python/mypy/issues?q=is%3Aissue+is%3Aopen+label%3Atopic-join-v-union) issue. Mypy's constraint solver uses joins rather than unions, so it evaluates `test(C1(), C2())` as `M` rather than `C1 | C2`.
